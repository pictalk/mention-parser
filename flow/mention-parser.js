// @flow
import { Result } from "safe-types";

type MapFunc<T, U> = (t: T) => U;

declare interface Tag {
  type: string;
  value: string;
}

declare type ErrorCode = {
  /**
   * A zero-valued error is not an error.
   * Currently unused.
   */
  None: 0,
  "0": "None",
  /**
   * Syntax error when a tag open/close token is used inside a tag
   * as a type or value.
   */
  InvalidToken: 1,
  "1": "InvalidToken",
  /**
   * Syntax error where the parser is trying to close the tag,
   * but does not receive the correct sequence of tokens.
   */
  UnclosedTag: 2,
  "2": "UnclosedTag",
};

declare interface ParseError {
  code:
    | $PropertyType<ErrorCode, "None">
    | $PropertyType<ErrorCode, "InvalidToken">
    | $PropertyType<ErrorCode, "UnclosedTag">;
  message: string;
}

declare function parse(message: string): Result<Tag[], ParseError>;

declare type TagToStringFunc = MapFunc<Tag, string>;
declare type TagToStringAsyncFunc = MapFunc<Tag, string | Promise<string>>;

declare function stringify(
  tags: Tag[],
  matcher?: {
    /**
     * Called when tag type not present in the matcher is found.
     */
    unknown?: TagToStringFunc,
    [tagType: string]: TagToStringFunc | void,
  },
): string;

declare function stringifyAsync(
  tags: Tag[],
  matcher?: {
    /**
     * Called when tag type not present in the matcher is found.
     */
    unknown?: TagToStringFunc,
    [tagType: string]: TagToStringAsyncFunc | void,
  },
): Promise<string>;
