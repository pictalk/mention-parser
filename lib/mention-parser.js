"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var parser_1 = require("./parser");
exports.ErrorCode = parser_1.ErrorCode;
exports.parse = parser_1.parse;
var stringer_1 = require("./stringer");
exports.stringify = stringer_1.stringify;
exports.stringifyAsync = stringer_1.stringifyAsync;
//# sourceMappingURL=mention-parser.js.map