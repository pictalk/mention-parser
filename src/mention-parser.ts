export { Tag } from "./shared";
export { ErrorCode, parse } from "./parser";
export {
  stringify,
  stringifyAsync,
  TagToStringFunc,
  TagToStringAsyncFunc,
  TagMatcher,
  TagMatcherAsync,
} from "./stringer";
