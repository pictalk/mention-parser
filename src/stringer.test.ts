import * as Tags from "./mention-parser";

interface User {
  id: string;
  fullname: string;
}

const users: { [name: string]: User } = {
  alex: {
    id: "37a8733e1718dc73d69f53ac",
    fullname: "Alex Regan",
  },
  frankie: {
    id: "9bfb2a90d20ef5eacf28bb2b",
    fullname: "Frankie Bagnardi",
  },
  mike: {
    id: "2903c8b677279552a8561956",
    fullname: "Mike Havrilla",
  },
  // x: {
  //   id: "f7429d93d90d65d49208d115",
  //   fullname: "",
  // },
  // x: {
  //   id: "ac09d7776de894209e1e70a1",
  //   fullname: "",
  // },
  // x: {
  //   id: "54d6b784b4bb6ba8893306ab",
  //   fullname: "",
  // },
};

const db: { [id: string]: User } = {};

Object.keys(users).reduce((db, k) => {
  let u = users[k];
  db[u.id] = u;
  return db;
}, db);

const sleep: (ms: number) => Promise<void> = ms =>
  new Promise(r => setTimeout(r, ms));

describe("Mention Stringer", () => {
  it("should work", () => {
    let r = Tags.parse(
      [
        `This should stringify a message with any set of {{hashtag:tags}}.`,
        `If I mention {{mention:${
          users.alex.id
        }}}, I should see his {{bold:name}}.`,
        `Maybe even some {{html:raw html}}?`,
      ].join(" "),
    );

    r.map_err(console.error);
    expect(r.is_ok()).toBe(true);
    expect(Tags.stringify(r.unwrap())).toMatchSnapshot("Default");

    expect(
      Tags.stringify(r.unwrap(), {
        text: t => t.value,
        mention: t => db[t.value].fullname,
        html: t => `<html>${t.value}</html>`,
      }),
    ).toMatchSnapshot("Custom");
  });

  it("should work with async resolvers", async () => {
    let r = Tags.parse(
      [
        `This should stringify an async message with any set of {{hashtag:tags}}.`,
        `If I mention {{mention:${
          users.frankie.id
        }}}, I should see his {{bold:name}}.`,
        `Maybe even some {{html:raw html}}?`,
      ].join(" "),
    );
    r.map_err(console.error);
    expect(r.is_ok()).toBe(true);

    let s = await Tags.stringifyAsync(r.unwrap(), {
      text: t => t.value,
      mention: async t => sleep(50).then(() => db[t.value].fullname),
      html: async t => `<html>${t.value}</html>`,
    });

    expect(s).toMatchSnapshot("Custom Async");
  });
});
