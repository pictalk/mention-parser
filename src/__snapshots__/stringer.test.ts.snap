// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mention Stringer should work with async resolvers: Custom Async 1`] = `"This should stringify an async message with any set of tags. If I mention Frankie Bagnardi, I should see his name. Maybe even some <html>raw html</html>?"`;

exports[`Mention Stringer should work: Custom 1`] = `"This should stringify a message with any set of tags. If I mention Alex Regan, I should see his name. Maybe even some <html>raw html</html>?"`;

exports[`Mention Stringer should work: Default 1`] = `"This should stringify a message with any set of tags. If I mention 37a8733e1718dc73d69f53ac, I should see his name. Maybe even some raw html?"`;
